gapis/api/service.proto

```diff
@@ -238,3 +238,117 @@ message CubemapLevel {
+  repeated MemoryAllocation allocations = 2;
+  // The index to get the constants names for MemoryAllocation `flags`.
+  // A value of -1 indicates no flag names should be fetched.
+  int32 allocation_flags_index = 3;
+  // The index to get the constants names for SparseImageBlock `aspects`.
+  // A value of -1 indicates no flag names should be fetched
+  int32 image_aspects_flags_index = 4;
+}
+
+// A single memory allocation
+message MemoryAllocation {
+  // The device this allocation was made on
+  uint64 device = 1;
+  // The memory type this allocation is from
+  uint32 memory_type = 2;
+  // Flags describing properties of the allocation, e.g. whether it's device
+  // local, whether it's host visible, etc.
+  uint32 flags = 3;
+  // Constant index for looking up the flag values with the path.ConstantSet endpoint
+  uint32 constant_set = 4;
+  // The API specific ID for this allocation
+  uint64 handle = 5;
+  // The user-readable name for this allocation
+  string name = 6;
+  // This size of the memory allocated
+  uint64 size = 7;
+  // The regions of host memory this allocation has been mapped into
+  MemoryMapping mapping = 8;
+  // All buffers/images bound to this memory allocation
+  repeated MemoryBinding bindings = 9;
+}
+
+// A mapping from part of a memory allocation into host memory
+message MemoryMapping {
+  // The size in bytes of the mapping
+  uint64 size = 1;
+  // The offset of the mapping into the source allocation, in bytes
+  uint64 offset = 2;
+  // The address on the host that this memory is bound to
+  uint64 mapped_address = 3;
+}
+
+// A buffer/image that is bound to part of a memory allocation
+message MemoryBinding {
+  // The API specific id for this object
+  uint64 handle = 1;
+  // The user-readable name for this binding
+  string name = 2;
+
+  // The size of the binding, in bytes
+  uint64 size = 3;
+  // The offset into the bound allocation, in bytes
+  uint64 offset = 4;
+
+  // Which type of object is using this memory, and any extra associated data
+  // with that.
+  oneof type {
+    // NOTE: bool is just here as a placeholder value
+    bool buffer = 5;
+    bool image = 6;
+    SparseImageBlock sparse_image_block = 7;
+    SparseImageMetadataMipTail sparse_image_metadata = 8;
+    SparseImageMetadataMipTail sparse_image_mip_tail = 9;
+    SparseBinding sparse_opaque_image_block = 10;
+    SparseBinding sparse_buffer_block = 11;
+  }
+}
+
+message SparseImageBlock {
+  // The offset into the image in the x-dimension
+  int32 x_offset = 1;
+  // The offset into the image in the y-dimension
+  int32 y_offset = 2;
+  // The width of the bound block
+  uint32 width = 3;
+  // The height of the bound block
+  uint32 height = 4;
+
+  // The mip level of the bound block
+  uint32 mip_level = 5;
+  // The array layer the bound block is in
+  uint32 array_layer = 6;
+
+  // The aspect(s) of the image this block is from (colour, depth, etc.)
+  uint32 aspects = 7;
```

Comment:
Can we use something other than a uint32_t here. Preferably a repeated field so we can have multiple aspects that have a proper name.

Reply:
Something like `repeated string` or do we still want to keep the values?  `repeated Aspect; message Aspect { uint32 value; string name; }`

Reply:
I think? this will work (we might be using an old version of proto but

enum AspectType {
    COLOR,
    DEPTH,
    STENCIL
}

repeated AspectType

Reply:
COLOR, DEPTH, STENCIL are pretty standard across all graphics APIs so I dont mind having them in the proto.

Reply:
Alright, metadata is technically also a vulkan aspect but I suppose it shouldn't show up except in `SparseImageMetadataMipTail`, where we've already distinguished it from regular mip tail.

Reply:
Sounds reasonable.